### PR TITLE
Refactoring  circle_of_trust.js and added Javascript Unit Tests for the same

### PR DESCRIPTION
**Pull request by fatimarafiqui**
This includes the following:
-  Refactoring of index.js to make it more modular.
-  Specs and test suites mocking methods through fixtures for circle_of_trust.js.
-  Block comments with each test suite name and description.
-  Separation of errorSpec.js from indexSpec.js 

selection_006

### CODE REVIEW

File: tests/jsTests/indexSpec.js [hunk 1]
```diff
@@ -1,43 +1,4 @@
 /**
- * Suite : Tests for showing error on incorrect field entry
- * Description : This is a test suite for showing error in case of an invalid field entry
- */
-describe("Tests for showing error on incorrect field entry", function() {
```
Comment (MehaKaushik) [inline tests/jsTests/indexSpec.js:5]: Why are these deleted?

Comment (fatimarafiqui) [reply tests/jsTests/indexSpec.js:5]: They are separated to errorSpec.js to enhance modularity. The corresponding js file is also  made cleaner. @MehaKaushik

File: tests/jsTests/circleOfTrustsSpec.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+					<li>fifth</li> \
+					<li>sixth</li> \
+					<li>seventh</li> \
+				</ul> \
+			</div>');
+		placeCircleOfTrustIcons();
+		jasmine.clock().tick(4000);
+	});
+
+	it("b", function() {
```
Comment (MehaKaushik) [inline tests/jsTests/circleOfTrustsSpec.js]: What is "b" here?

File: tests/jsTests/circleOfTrustsSpec.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+			<div class="actions">\
+				<div class="ui cancel green button">Close</div>\
+			</div>\
+		</div>');
+		$('.ui.modal').hide();
+
+		// mock modal show call
+		spyOn($.fn, "modal").and.returnValue($('.ui.modal').show());
+		showResponse($('element'), { response: true, message: msg });
+		jasmine.clock().tick(4000);
```
Comment (MehaKaushik) [inline tests/jsTests/circleOfTrustsSpec.js]: Write a small comment here, why you are using clock feature.

**Review by MehaKaushik — DISMISSED** [review]
Please address the minor comments, before merging it.

File: tests/jsTests/circleOfTrustsSpec.js (hunk 2 above)
Comment (fatimarafiqui) [reply tests/jsTests/circleOfTrustsSpec.js]: Fixed here https://github.com/systers/FirstAide-web/pull/264/commits/68637a7ce80b3d5487de124ba90dad59ff402280

File: tests/jsTests/circleOfTrustsSpec.js (hunk 3 above)
Comment (fatimarafiqui) [reply tests/jsTests/circleOfTrustsSpec.js]: Done here https://github.com/systers/FirstAide-web/pull/264/commits/68637a7ce80b3d5487de124ba90dad59ff402280

File: tests/jsTests/errorSpec.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * Suite : Tests for hiding error on correct field entry
+ * Description : This is a test suite for hiding error in case of an valid field entry
+ */
+describe("Tests for hiding error on correct field entry", function() {
+	var msg = 'This is a test message for not showing error';
+	beforeEach(function() {
+		setFixtures('<div class="field"> \
+				<div class="ui red pointing">' + msg + '</div> \
+			</div>');
+		hideError($('.field'));
```
Comment (npari) [inline tests/jsTests/errorSpec.js]: @fatimarafiqui In this test, you should create the error and make it visible in the setFixtures method in order to really test if hideError is working as expected.

File: tests/jsTests/circleOfTrustsSpec.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		});
+		spyOn($.fn, "fadeIn").and.callFake(function() {
+			$(this).show();
+		});
+        toggleCircleOfTrustSections('help');
+		// creating delay of 400ms
+		jasmine.clock().tick(4000);
+	});
+
+	it("get help button clicked", function() {
```
Comment (npari) [inline tests/jsTests/circleOfTrustsSpec.js]: The line "get help button clicked" does not describe what is expected in the test. I see that the test expects edit option to be hidden and comrade action section to be visible. Let's write something else that better depicts the test.

File: tests/jsTests/circleOfTrustsSpec.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+		expect($('.ui.modal')).toBeVisible();
+	});
+});
+
+/**
+ * Suite : Tests for failed response modal
+ * Description : This is a test suite for a failed response modal
+ * in case of a failure
+ */
+describe("Tests for failed response modal", function() {
```
Comment (npari) [inline tests/jsTests/circleOfTrustsSpec.js:44]: Lets add a test to check if showResponse handles the parameters as expected when the 'response' parameter is undefined.

File: includes/page_top.php [hunk 7]
```diff
@@ -3,6 +3,8 @@
     $page_request = $_GET['page_request'] ?? \FirstAide\Router::INDEX;
     $query = $_GET['query'] ?? '';
     $page = FirstAide\Router::getPage($UserObj, $page_request, $query);
+    
+    $page['javascripts'][] = 'error.js';
```
Comment (npari) [inline includes/page_top.php:8]:
```
<title><?php` echo isset($_page['title']) ? $_page['title'].' | ' : ''  ?>FirstAide, Peacecorps</title>
````
Let's add a default title in case $_page['title] is not set rather than keeping it empty.

**Review by npari — CHANGES_REQUESTED** [review]

File: js/circle_of_trust.js [hunk 8]
```diff
@@ -1,14 +1,3 @@
...
-    setTimeout(function() {ele.find('.ui.red').hide();}, 7000);
-}
-function hideError(ele) {
-    ele.removeClass('error');
-    ele.find('.ui.red.pointing').hide();
-}
-
 function showResponse(response, pageReload) {
     if (typeof pageReload === 'undefined') { pageReload = true; }
     if (!response.response) {
```
Comment (Anjali05) [inline js/circle_of_trust.js]: Put some proper message inside this if.

**Review by Anjali05 — APPROVED** [review]

**Review by Anjali05 — APPROVED** [review]

**Review by Anjali05 — APPROVED** [review]

**Review by Anjali05 — DISMISSED** [review]

File: tests/jsTests/errorSpec.js (hunk 4 above)
Comment (fatimarafiqui) [reply tests/jsTests/errorSpec.js]: @npari In HTML every dome element is visible by default, hence this fixture is creating the error and making it visible simultaneously. So, it is really testing the working of hideError as expected.

File: includes/page_top.php (hunk 7 above)
Comment (fatimarafiqui) [reply includes/page_top.php:8]: @npari The $_page['title] is being set by Router.php for every condition and for each page. But in other cases, the default title is the name of the website i.e FirstAide, Peacecorps. I have further improved this here -  https://github.com/systers/FirstAide-web/pull/264/commits/02921def05799f50a0595cd83ff76eaf1dbea433

Done.

File: tests/jsTests/circleOfTrustsSpec.js (hunk 6 above)
Comment (fatimarafiqui) [reply tests/jsTests/circleOfTrustsSpec.js:44]: Done here https://github.com/systers/FirstAide-web/pull/264/commits/a1520b54287ea7c45e785b2100053401cb49cb4a Completely rewritten the function and added the tests.

File: tests/jsTests/circleOfTrustsSpec.js (hunk 5 above)
Comment (fatimarafiqui) [reply tests/jsTests/circleOfTrustsSpec.js]: Done https://github.com/systers/FirstAide-web/pull/264/commits/b789dd75e00a3e654cf73000770ce480ccae7130

File: js/circle_of_trust.js (hunk 8 above)
Comment (fatimarafiqui) [reply js/circle_of_trust.js]: Done. Updated the complete method.

File: tests/jsTests/errorSpec.js (hunk 4 above)
Comment (npari) [reply tests/jsTests/errorSpec.js]: @fatimarafiqui 
https://github.com/systers/FirstAide-web/pull/264/files/68637a7ce80b3d5487de124ba90dad59ff402280#diff-5225fdc1ab35e654deeac6f8b75ab2acR24 Thats the hideError method that you have written. The method removes 'error' class. In your setFixtures, you need to set the 'error' class, then call `hideError` and then add the assert to see if the `'error'` class is removed from the html.  In this case, the 'error' class is not present even before calling the hideError method, so it true by default.

**Review by npari — CHANGES_REQUESTED** [review]

**Comment by fatimarafiqui** [comment]
@npari 
Code A (from the link you gave):
```
function test(pageReload = true) {
```

Code B (from the link you gave):
```
pageReload = (typeof pageReload !== 'undefined') ?  pageReload : true;
```

Code C ( [Here in code](https://github.com/fatimarafiqui/FirstAide-web/blob/68637a7ce80b3d5487de124ba90dad59ff402280/js/circle_of_trust.js#L8) ):
```
 if (typeof pageReload === 'undefined') { pageReload = true; } 
```

All of them do the same thing. Code B and C supports pre ES2015 dependencies (see references and link you have shared). Hence it is done to support backward compatibilities. This would be helpful in the future development of the code. Moreover, eliminating if conditions will not give any functional improvement in JS.

Reference: [Default parameters JS](https://stackoverflow.com/questions/894860/set-a-default-parameter-value-for-a-javascript-function)

**Review by Anjali05 — APPROVED** [review]

**Review by npari — APPROVED** [review]

**Review by npari — APPROVED** [review]
